content/article/my-perl-wishlist-invariant-sigils-part-1.md

```diff
@@ -0,0 +1,181 @@
+----------------------
+
+Slices such as `@z[0,1]` and `%z{qw(hello there)}` return multiple
+values from an array or hash.  If sigils `@` and `%` are no longer
+available for slicing, we need an alternative.
+The Perl community currently provides two models: postderef syntax and
+postfix adverbs.
+
+Perl 5.20+ support
+[postfix dereferencing](https://www.effectiveperlprogramming.com/2014/09/use-postfix-dereferencing/)
+("postderef"), which gives us one option.  Postderef separates the name
+from the slice:
+
+```perl
+# Valid Perl 5.20+
+$hashref->{a};      # Scalar, element at index "a" of the hash pointed to by $hashref
+$hashref->@{a};     # List including the "a" element of the hash pointed to by $hashref
+$hashref->%{a};     # List including the key "a" and the "a" element of the hash pointed to by $hashref
+```
+
+The type of slice comes after the reference, instead of as a sigil
+before the reference.  With non-references, that would give us slices
+such as `@z@[1,2,3]` or `%hash%{a}`.
+
+Raku gives us another option: "adverbs" such as
+[`:kv`](https://docs.perl6.org/language/subscripts#:kv).  For example:
+
+```perl6
+# Valid Raku
+%hash{"a"}          # Single value, element at index "a" of %hash
+%hash{"a"}:v;       # The same --- just the value
+%hash{"a"}:kv;      # The list including key "a" and the value of the "a" element of %hash
+```
+
+The adverb (e.g., `:kv`) goes in postfix position, immediately
+after the brackets or braces.  Following this model,
+slices would look like `@z[1,2,3]:l` or `%hash{a}:kv`.  (I propose `:l`
+instead of `:v` to reduce confusion, since Raku's `:v` can return a
+scalar or a list.)
+
+So, the choices I see are (postderef-inspired / Raku-inspired):
+
+| What you want           | No subscript                                    | `[ ]` access                                                                 | `{ }` access                                                                          |
+|-------------------------|-------------------------------------------------|------------------------------------------------------------------------------|---------------------------------------------------------------------------------------|
+| Scalar                  | `$z`: a scalar, i.e., a single value            | `@z[0]`: A single value from an array                                        | `%z{0}`: a single value from a hash                                                   |
+| List of values          | `@z`: An array, i.e., a list of value(s)        | `@z@[0]` / `@z[0]:l`: a list including the first element of `@z`             | `%z@{0}` / `%z{0}:l`: A list including the `"0"` element of `%z`                        |
+| List of key/value pairs | `%z`: A hash, i.e., a list of key/value pair(s) | `@z%[0]` / `@z[0]:kv`: a list of the index `0` and the first element of `@z` | `%z%{0}` / `%z{0}:kv`: a list of the key `"0"` and the value of element `"0"` of `%z` |
+
+You can't always get what you want
```

Comment:
I love how you dug into the parser and what's possible here.

Reply:
Just wait until part 2 :D